### PR TITLE
Delete Comment Error Checking

### PR DESCRIPTION
**Pull request by lxcyqx**
DeleteCommentServlet - do not allow people to delete comments unless it is their own comment

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java [hunk 1]
```diff
@@ -23,10 +25,21 @@
...
-    long id = Long.parseLong(request.getParameter("id"));
-
-    Key commentEntityKey = KeyFactory.createKey("Comment", id);
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.delete(commentEntityKey);
+    UserService userService = UserServiceFactory.getUserService();
+    String commentEmail = request.getParameter("email");
+    String userEmail = request.getParameter("user");
+    //if user did not make the comment or if user is not logged in 
+    if (!commentEmail.equals(userEmail) || !userService.isUserLoggedIn()){
```
Comment (g-asweet) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:32]: This is excellent, you're doing server-side validation to ensure that nobody is pretending to be somebody else or hit this endpoint independently. This is true use of User Authentication. Nice work!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java [hunk 2]
```diff
@@ -23,10 +25,21 @@
...
+      System.err.println("ERROR: Cannot delete comment.");
+      return;
+    }
+    try {
+      long id = Long.parseLong(request.getParameter("id"));
+      Key commentEntityKey = KeyFactory.createKey("Comment", id);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.delete(commentEntityKey);
```
Comment (g-asweet) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:40]: When dealing with `try`/`catch`, try to keep as little logic as possible within the try statement. In this case I could recommend one of two fixes:

- Implement a `getCommentId(WHATEVER_PARAMS...)` which handles the try catch and returns a long with -1 or 0 when its invalid
- OR if you still want this try catch in this method:
```java
long id;
try {
  id = Long.parseLong(request.getParameter("id"));
} catch (NumberFormatException e){
  System.err.println("ERROR: Failed to parse to long.");
  // Invalid id, don't continue
  return;
}

...
```

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -178,9 +182,9 @@ function addComment() {
   document.getElementById("comment-box").value = "";
   fetch(
```
Comment (g-asweet) [inline portfolio/src/main/webapp/script.js:183]: Consider using string formatting, i.e.

```javascript
const five = 5;
const ten = 10;
console.log(`Fifteen is ${five + ten} and not ${2 * five + ten}.`);
```

**Comment by lxcyqx** [comment]
In DataServlet, changed the logic so that getting comments only requires one fetch request instead of the initial two (the second of which was used to determine whether current page was last page of comments). To do so, I created a CommentsInfo class, which contains an ArrayList of Comment objects (the comments that are to be displayed on the current page) and an integer indicating the number of comments on the next page. If this number is 0, that means current page is the last page of comments, for which case the next button is disabled

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java [hunk 4]
```diff
@@ -23,8 +25,22 @@
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (g-asweet) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:27]: [Optional] I'm curious now that you've gotten me thinking about page counts and edge cases, what happens if you're on page 3, it has only one comment, and you delete it? Do you have a means of reloading the comments and getting the user to page 2?

Comment (lxcyqx) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:27]: Hmm just checked that right now. Looks like I forgot to account for that edge case. I can work on that right now

**Pull request merged**